### PR TITLE
Refactored inventory page's sorting to use Utils.Sorter

### PR DESCRIPTION
**Pull request by jeudy100**
**System Name**
image
image

**Last Check In**
image
image

**Status**
image
image

### CODE REVIEW

**Comment by jeudy100** [comment]
fixes #309

File: app/js/states/inventory/inventory.controller.js [hunk 1]
```diff
@@ -70,8 +70,14 @@ function InventoryCtrl(
...
-    $scope.predicate = $location.search().sort_field || 'toString';
-    $scope.reverse = $location.search().sort_dir === 'DESC';
+    InventoryService.setSortField($location.search().sort_field || 'toString');
+    InventoryService
+        .setSortDirection($location.search().sort_dir);
+
+    $scope.sorter = new Utils.Sorter({
+        predicate: InventoryService.getSortField(),
+        reverse: booleanSortDirection()
+    }, order);
```
Comment (Kinlaw) [inline app/js/states/inventory/inventory.controller.js]: We should get rid of references to InventoryService sorting fields and use a default predicate stored in inventory.controller.js if there is no url param

File: app/js/states/inventory/inventory.controller.js [hunk 2]
```diff
@@ -298,41 +304,53 @@ function InventoryCtrl(
...
-                '-' + $scope.predicate :
-                $scope.predicate)]);
+            ($scope.sorter.reverse ?
+                '-' + $scope.sorter.predicate :
+                $scope.sorter.predicate)]);
+    }
+
+    function booleanSortDirection() {
+        return InventoryService.getSortDirection() === 'DESC';
     }
```
Comment (Kinlaw) [inline app/js/states/inventory/inventory.controller.js]: We should be able to get rid of this function.

File: app/js/services/utils.service.js [hunk 3]
```diff
@@ -249,22 +249,36 @@ function Utils($filter, $rootScope, Events) {
...
+                implicitOrder.changeReverse || false;
+
+            // disables calling of callback function.
+            this.disableCallback =
+                implicitOrder.disableCallback || false;
         } else {
             this.predicate = '';
             this.reverse = false;
+            this.changeReverse = false;
+            this.disableCallback = false;
```
Comment (Kinlaw) [inline app/js/services/utils.service.js]: Why are we controlling the callback in the sort utility?

File: app/js/services/utils.service.js [hunk 4]
```diff
@@ -249,22 +249,36 @@ function Utils($filter, $rootScope, Events) {
...
         }
 
         this.cb = cb;
     };
 
     utils.Sorter.prototype.sort = function (name) {
-        this.reverse = (this.predicate === name) ? !this.reverse : false;
+        this.reverse = (this.predicate === name) ? !this.reverse : this.changeReverse;
```
Comment (Kinlaw) [inline app/js/services/utils.service.js]: why do we need two different reverse variables?

File: app/js/states/inventory/inventory.controller.js [hunk 5]
```diff
@@ -298,40 +300,40 @@ function InventoryCtrl(
...
 
         // if we have the full inventory list then use local sorting
         if ($scope.systems.length === InventoryService.getTotal()) {
-            order();
+            if ($scope.sorter.disableCallback) {
+                $scope.sorter.disableCallback = false;
+            }
+
+            $scope.sorter.sort(column);
             $scope.loading = false;
```
Comment (Kinlaw) [inline app/js/states/inventory/inventory.controller.js]: let's just get rid of the local sorting for now. It's not working anyway. If I get all the systems and sort, I still see an api call going out. This should remove the need for a disableCallback flag.

File: app/js/services/utils.service.js (hunk 3 above)
Comment (Kinlaw) [reply app/js/services/utils.service.js]: See comments below, we should be able to get rid of this.

File: app/js/states/inventory/inventory.jade [hunk 6]
```diff
@@ -49,10 +49,10 @@
...
-              th.sortable(ng-class="predicate | sortClass:'toString':reverse", ng-click="sort('toString')", translate) System Name
-              th.fit-content.min.sortable(ng-class="predicate | sortClass:'last_check_in':!reverse", ng-click="sort('last_check_in')", translate) Last Check In
-              th.min.sortable(ng-class="predicate | sortClass:'report_count':reverse", ng-click="sort('report_count')", translate) Status
+              th.fit-content.min.sortable(ng-class="sorter.predicate | sortClass:'system_type_id':sorter.reverse", ng-click="sort('system_type_id')", translate) System Type
+              th.sortable(ng-class="sorter.predicate | sortClass:'toString':sorter.reverse", ng-click="sort('toString')", translate) System Name
+              th.fit-content.min.sortable(ng-class="sorter.predicate | sortClass:'last_check_in':sorter.reverse", ng-click="sort('last_check_in')", translate) Last Check In
+              th.min.sortable(ng-class="sorter.predicate | sortClass:'report_count':sorter.reverse", ng-click="sort('report_count')", translate) Status
```
Comment (Kinlaw) [inline app/js/states/inventory/inventory.jade]: we should be calling sorter.sort in the ng-click which will set reverse, and predicate. Then the callback is called which in this case is set to order(). Order should then reset the scope and call getData().

File: app/js/states/inventory/inventory.controller.js [hunk 7]
```diff
@@ -298,40 +300,40 @@ function InventoryCtrl(
     function order() {
         $scope.systems = $filter('orderBy')($scope.systems,
             ['_type',
-            ($scope.reverse ?
-                '-' + $scope.predicate :
-                $scope.predicate)]);
+            ($scope.sorter.reverse ?
+                '-' + $scope.sorter.predicate :
+                $scope.sorter.predicate)]);
```
Comment (Kinlaw) [inline app/js/states/inventory/inventory.controller.js]: $filter is a local sort. It depends on us having the complete systems data set. Inventory uses the api for paging so we only have one page at a time. We should call getData here.

File: app/js/services/utils.service.js [hunk 8]
```diff
@@ -249,22 +249,36 @@ function Utils($filter, $rootScope, Events) {
...
         this.predicate = name;
-        if (this.cb) {
+        if (this.cb && !this.disableCallback) {
             this.cb();
         }
     };
 
+    utils.Sorter.prototype.getSortDirection = function () {
+        return this.reverse ? 'DESC' : 'ASC';
+    };
```
Comment (Kinlaw) [inline app/js/services/utils.service.js]: Good addition!

**Comment by Kinlaw** [comment]
@jeudy100 also, can you remove the sorting variables from InventoryService? I think there are a few references to them in the app but they shouldn't be needed. If anything does need to know the sort dir or sort field then they should be using the url params. Each time we sort the data set should refresh because we are querying the api. You can listen on $scope.systems to determine if those variables refresh.

File: app/js/states/inventory/inventory.controller.js [hunk 9]
```diff
@@ -209,12 +220,6 @@ function InventoryCtrl(
...
         FilterService.parseBrowserQueryParams();
-        if (params.sort_field) {
-            InventoryService.setSortField(params.sort_field);
-            if (params.sort_dir) {
-                InventoryService.setSortDirection(params.sort_dir);
-            }
-        }
     }
 
     parseBrowserQueryParams();
```
Comment (Kinlaw) [inline app/js/states/inventory/inventory.controller.js]: We can probably get rid of the local function parseBrowserQueryParams and replace this line with FilterService.parseBrowserQueryParams();

File: app/js/components/topbar/alerts/alerts.directive.js [hunk 10]
```diff
@@ -74,6 +70,8 @@ function alertsCtrl($rootScope,
         }
     }
 
+    $scope.goToInventory = goToInventory;
```
Comment (Kinlaw) [inline app/js/components/topbar/alerts/alerts.directive.js]: is this being used?

Comment (jeudy100) [reply app/js/components/topbar/alerts/alerts.directive.js]: no its not

**Pull request merged**